c/certificateService.c

```diff
@@ -118,36 +119,85 @@ static void respondWithBadRequest(HttpResponse *response) {
 
     jsonStart(p);
     {
-      jsonAddString(p, "error", "The length of the certificate is longer than 4096 bytes");
+      jsonAddString(p, "error", value);
     }
     jsonEnd(p);
 
     finishResponse(response);
 }
 
+
 static int serveMappingService(HttpService *service, HttpResponse *response)
 {
   HttpRequest *request = response->request;
 
   if (!strcmp(request->method, methodPOST))
   {
-    RUsermapParamList *userMapCertificateStructure 
-      = (RUsermapParamList*)safeMalloc31(sizeof(RUsermapParamList),"RUsermapParamList");
-    memset(userMapCertificateStructure, 0, sizeof(RUsermapParamList));
+    RUsermapParamList userMapStructure = {0};
 
-    if(request->contentLength > sizeof(userMapCertificateStructure->certificate) || request->contentLength < 0) {
-      respondWithBadRequest(response);
-      return 0;
+    memset(&userMapStructure, 0, sizeof(RUsermapParamList));
+    int urlLength = strlen(request->uri);
+    if(urlLength < 0 || urlLength > 64) {
+        respondWithBadRequest(response, "URL length is not in range from 0 to 64 bytes.");
+        return 0;
+    }
+    char translatedURL[urlLength];
+    memcpy(translatedURL, request->uri, strlen(request->uri));
+    a2e(translatedURL, sizeof(translatedURL));
+    char *found = strstr(translatedURL,"x509");
+
+    if(found != NULL) {
+    //  Certificate to user mapping
+        if(request->contentLength > sizeof(userMapStructure.certificate) || request->contentLength < 0) {
+          respondWithBadRequest(response, "The length of the certificate is longer than 4096 bytes");
+          return 0;
+        }
+
+        userMapStructure.certificateLength = request->contentLength;
+        memset(userMapStructure.certificate, 0, request->contentLength);
+        memcpy(userMapStructure.certificate, request->contentBody, request->contentLength);
+
+        userMapStructure.functionCode = MAP_CERTIFICATE_TO_USERNAME;
+    } else {
+    //    Distinguished ID to user mapping
+        char *bodyNativeEncoding = copyStringToNative(request->slh, request->contentBody, strlen(request->contentBody));
+        char errorBuffer[2048];
+        Json *body = jsonParseUnterminatedString(request->slh, bodyNativeEncoding, strlen(bodyNativeEncoding), errorBuffer, sizeof(errorBuffer));
```

Comment:
Even the function name says `UnterminatedString` and at the same time `strlen` is used.